Added task 1

Dx18 opened the pull request.

Mrkol committed:
Initial commit

Mrkol committed:
Initial migration from vk_graphics_basic

Commit:
Added task 1

Commit:
Added task 1

Commit:
Fixed the BRDF

Mrkol commented on tasks/task1/CMakeLists.txt:
```diff
@@ -0,0 +1,27 @@
+
```
Не надо на папки task1/task2/task3 разбивать, наша цель сделать один единый проект в котором красивый рендер, а не кучу маленьких игрушечных тасок. В частности, лучше подробно писать в PR что ты сделал, а не просто task1 (что вообще за task1? у нас 1 таск на весь семестр, сделать красивый рендер =))

Mrkol commented on common/render_utils/CMakeLists.txt:
```diff
@@ -1,5 +1,5 @@
-
-add_library(render_utils QuadRenderer.cpp)
+add_library(render_utils
+  LineRenderer.cpp QuadRenderer.cpp)
```
Забыл glm::glm подрубить, без этого у меня не компилируется (и странно, что компилируется у тебя)

Commit:
Removed command buffer submission and made the buffer readonly

Commit:
Fixed image layout transitions

Mrkol commented on common/scene/SceneMaterials.hpp:
```diff
@@ -0,0 +1,51 @@
+  std::optional<std::size_t> normalTexture;
+  std::optional<std::size_t> occlusionTexture;
+
+  std::optional<std::size_t> emissiveTexture;
```
Рекомендую везде для айдишников каких-то сущностей использовать не std::size_t, а `enum class TextureId : uint32_t {}`. Тогда не возможно случайно что-то не то куда-то не туда пихнуть, потому что всё типизированное и нет неявных кастов. Ну и `uint32_t` вместо `size_t` потому что больше 2 млн текстур никому нигде не понадобится, а размер структуры material в памяти уменьшиться значительно, что положительно с кажется на кешах. Конечно лучше мерить, но в качестве дефолта лучше всё таки uint32_t использовать для чего-то что в памяти лежит а не только через аргументы кидается туда-сюда

Commit:
Fixed `emissiveFactor` assignment

Commit:
Made some fixes

1. Removed `colorIn` from BRDF
2. Optimized the G-Buffer images formats
3. Doing one pass for all the lights instead of one pass for each
   light type

Commit:
Removed quad rendering

Dx18 commented:
Скриншоты:

![screenshots](https://github.com/user-attachments/assets/4717e9ab-eae4-4e1f-b677-c0e68b7fc061)

Commit:
Made image/sampler/texture/material indices type-safe

Commit:
Removed tangents, compressed push constants

Commit:
Fixed synchronization, removed deprecated extension

Commit:
Removed LFS

Commit:
Removed LFS

Commit:
Fixed specular issues connected to normal precision

Commit:
Cleanup and fixed after rebase

Commit:
Added more missing files

Commit:
Changed etna tag

Commit:
Added intensified light color calculation

Commit:
Fixed constant naming in `distributionGgx`

Commit:
Fixed shader formatting and material type setting

Dx18 commented:
Новые скриншоты:
![screenshots](https://github.com/user-attachments/assets/5a6719fa-1513-4e4c-9830-d6ccd41c467f)

Commit:
An attempt to make diffuse baking

Commit:
Given up on diffuse baking and fixed (?) the spherical harmonics

Commit:
Removed explicit device selection

Commit:
Added missing cubemaps

Dx18 commented:
Скриншоты (для диффуза):
![screenshots_diffuse](https://github.com/user-attachments/assets/bed7775a-3f54-441a-bce4-53c0bf377368)

Commit:
Approximating actual diffuse with spherical harmonics now

s59dpodk commented:
![screenshots_diffuse](https://github.com/user-attachments/assets/bacdd23d-06da-4b65-ac10-c07b671007f5)

Dx18 commented:
![screenshots_diffuse](https://github.com/user-attachments/assets/b0022836-41d2-4f3f-8857-9aacf585144f)
